### DISCUSSION TITLE
Running in the background

### DISCUSSION THREAD

**Discussion by amadensor** [miguelgrinberg/microdot#237] [Q&A] [answered]
Is it possible to have the web server running in the background and have other tasks also executing?   I want to build a project that has a web interface to do the configuration, but uses GPIO pins and catches actions from physical buttons.  This would obviously be a separate loop executing.

**Comment by amadensor** [comment]
I made it do the thing, but there might be a better way:

```
async def main():
    gpioloop=asyncio.create_task( do_things())
    web=asyncio.create_task( app.start_server(port=80, debug=True))
    await gpioloop
    await web

async def do_things():
    while True:
        print("Do More")
        await asyncio.sleep(1)

asyncio.run(main())
```

**Reply by miguelgrinberg** [reply, marked as answer]
This is fine, but given that you do not have anything to do in the main asyncio task I would suggest you run one of the two in a background task and the other in the main one.

**Reply by amadensor** [reply]
Hopefully this helps other also.

**Reply by pwittich** [reply]
> This is fine, but given that you do not have anything to do in the main asyncio task I would suggest you run one of the two in a background task and the other in the main one.

Can I ask for more information about your response? I'm running into a similar issue. When you say 'background task' do you mean something like
```
async def main():
    _thread.start_new_thread(do_something,())
    web=asyncio.create_task( app.start_server(port=80, debug=True))
    await web
```

That is what I am trying to do, so far without a lot of luck. 

(I can start a new thread  but this one is quite what I want to do. What I'm finding is that the application stalls when a request comes in, for reasons I am trying to figure out.)

**Reply by miguelgrinberg** [reply]
Thread support in microcontroller boards is very limited or many times not available. And besides, Microdot is an async framework, so threads cannot be used, even if they are available.

To start a job in the background you use `asyncio.create_task`. For example:

```python
async def main():
    web = asyncio.create_task( app.start_server(port=80, debug=True))
    while True:
        # do something else here, while the web server runs in the background
    await web  # wait for the web server to end
```

**Reply by madflier** [reply]
> ```python
> async def main():
>     web = asyncio.create_task( app.start_server(port=80, debug=True))
>     while True:
>         # do something else here, while the web server runs in the background
>     await web  # wait for the web server to end
> ```

<naive poster mode>

Why the `await web` line here? Presumably it is never reached by the code.

</naive poster mode>

**Reply by miguelgrinberg** [reply]
If your web server does not exit gracefully then you can omit the await at the bottom. But a static analysis tool may complain that you have a task that is not awaited.
